### PR TITLE
Add TCF convolutional op with bias addition

### PR DESCRIPTION
**Pull request by ajarthurs**

### CODE REVIEW

**Review by silvasean — CHANGES_REQUESTED** [review]

File: test/npcomp-run-mlir/conv_2d_nchw_bias.mlir [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// RUN: npcomp-run-mlir %s \
```
[the remark this answers is not in the document]
Comment (ajarthurs) [reply test/npcomp-run-mlir/conv_2d_nchw_bias.mlir:1]: OK. I'll do that.

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 2]
```diff
@@ -95,4 +95,24 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
...
+  let description = [{
+    Performs 2-D convolution plus bias addition.
+
+    The tensors have dimensions:
+    - in:     [N, C, H, W]
+    - filter: [F, C, KH, KW]
+    - bias:   [N, F, H, W]
+    - result: [N, F, H, W]
+
+    If the `K` dimension mismatches between the operands, this op aborts the
```
[the remark this answers is not in the document]
Comment (ajarthurs) [reply include/npcomp/Dialect/TCF/IR/TCFOps.td:109]: Ah! That sentence describes `matmul`'s inside-dimension constraint, which is obviously incorrect for this op. I'll reword it.

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 3]
```diff
@@ -95,4 +95,24 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
   let assemblyFormat = "$lhs `,` $rhs attr-dict `:` functional-type(operands, results)";
 }
 
+def TCF_ConvNCHWBiasOp : TCF_Op<"conv_2d_nchw_bias"> {
+  let summary = "2-D convolution with bias";
+  let description = [{
+    Performs 2-D convolution plus bias addition.
```
[the remark this answers is not in the document]
Comment (ajarthurs) [reply include/npcomp/Dialect/TCF/IR/TCFOps.td:101]: OK. I'll add a commit to remove bias-addition.

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp [hunk 4]
```diff
@@ -76,6 +86,65 @@ class ConvertMatmul : public OpRewritePattern<tcf::MatmulOp> {
...
+    Value matchingFC =
+        rewriter.create<CmpIOp>(op.getLoc(), CmpIPredicate::eq, inputC, filterF);
+    Value validFilterH =
+        rewriter.create<CmpIOp>(op.getLoc(), CmpIPredicate::uge, inputH, filterKH);
+    Value validFilterW =
+        rewriter.create<CmpIOp>(op.getLoc(), CmpIPredicate::uge, inputW, filterKW);
+    Value witnessC = rewriter.create<shape::CstrRequireOp>(
+        op.getLoc(), matchingC, "input and filter channels must be equal");
+    Value witnessFC = rewriter.create<shape::CstrRequireOp>(
+        op.getLoc(), matchingFC, "input channels and filter F-dimension must be equal");
```
Comment (silvasean) [inline lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:115]: I don't think "C" and "F" is commonly used terminology when describing the filter tensor. Usually I hear Zin/Zout terminology or input channels / output channels. And I don't understand this check either. The number of output channels is not constrained.

File: test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// RUN: not npcomp-run-mlir %s \
+// RUN:   -invoke conv_2d_nchw \
+// RUN:   -arg-value="dense<[[[[1.0, 0.0], [1.0, 1.0]]]]> : tensor<1x1x2x2xf32>" \
+// RUN:   -arg-value="dense<[[[[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]]]]> : tensor<1x1x3x2xf32>" \
+// RUN:   -shared-libs=%npcomp_runtime_shlib 2>&1 \
+// RUN:   | FileCheck %s --check-prefix=HEIGHT
+
+// RUN: not npcomp-run-mlir %s \
+// RUN:   -invoke conv_2d_nchw \
+// RUN:   -arg-value="dense<[[[[1.0, 0.0], [1.0, 1.0]]]]> : tensor<1x1x2x2xf32>" \
```
Comment (silvasean) [inline test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir:24]: for the purpose of this test, you can just use `dense<0.0> : tensor<...>`, since we don't care about the tensor contents.

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 6]
```diff
@@ -95,4 +95,25 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
...
+  let description = [{
+    Performs 2-D convolution.
+
+    The tensors have dimensions:
+    - in:     [N, C, H, W]
+    - filter: [F, C, KH, KW]
+    - result: [N, F, H, W]
+
+    The tensors must met the following conditions; otherwise, this op aborts the program.
+    - F equals C
```
Comment (silvasean) [inline include/npcomp/Dialect/TCF/IR/TCFOps.td:109]: F == C is not required.

**Review by silvasean — CHANGES_REQUESTED** [review]

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 4 above)
Comment (ajarthurs) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:115]: I borrowed MLIR-Linalg's naming convention for the tensor dimensions in its conv-ops, illustrated in the following code snippet from said dialect. If that is not acceptable, what exactly do you suggest for the dimension names?
```
# llvm-project/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOpsSpec.tc
ods_def<ConvNCHWOp>:
def conv_2d_nchw(I: f32(N, C, H, W), K: f32(F, C, KH, KW)) -> (O: f32(N, F, H, W)) {
  O(n, f, h, w) = std_addf<kh, kw>(std_mulf(
    I(n, c, h + kh, w + kw), K(f, c, kh, kw)));
}
```
And, I neglected to constrain the output tensor and just focused on the two inputs, so I need to add constraints for the output tensor as well, correct? Or no?

File: test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir (hunk 5 above)
Comment (ajarthurs) [reply test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir:24]: OK. I didn't pay attention to the tensor contents since it's strictly a test on shapes, but I'll apply your suggestion.

Comment (ajarthurs) [reply test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir:24]: For the most part, I just copied-and-pasted the `invalid-matmul.mlir` test.

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 4 above)
Comment (silvasean) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:115]: Output tensor shape is calculated here (and we assume that it is a correct calculation, otherwise there is a bug in our code), so we don't need to check the output shape.

For now, we can keep the names as-is, but let's make sure the error messages say "output channels" or "input channels" when referring to the filter dimensions.

File: test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir (hunk 5 above)
Comment (silvasean) [reply test/npcomp-run-mlir/invalid-conv_2d_nchw.mlir:24]: Oops, I should have followed my own advice there ;) Thanks for catching that!

File: include/npcomp/Dialect/TCF/IR/TCFOps.td (hunk 6 above)
Comment (ajarthurs) [reply include/npcomp/Dialect/TCF/IR/TCFOps.td:109]: OK.

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 4 above)
Comment (ajarthurs) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:115]: I have some confusion about output shape calculation. Are you pointing out that the filter's channel size is transferred to the output channel size internally? The present TCF->Linalg conversion code doesn't cover the output shape. Regardless, it sounds like no changes are needed in scope of this conversation.

Comment (silvasean) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:115]: The filter has two channel sizes 1) the number of input channels and 2) the number of output channels.

The TCF->Linalg conversion does cover the output shape. That is what the bypassResultShapes call is doing.

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp [hunk 7]
```diff
@@ -32,6 +32,16 @@ static SmallVector<Value, 6> bypassResultShapes(Operation *op,
         op->getLoc(), ValueRange({lhsRows, rhsCols}));
     return {shape};
   }
+  // TODO: This only supports the NCHW data format. Consider other formats and lower ranks.
+  if (auto conv2dNCHW = dyn_cast<tcf::ConvNCHWOp>(op)) {
```
Comment (silvasean) [inline lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:36]: This is not correct. The general formula is here: https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html

image

In this case, padding=0, dilation=1, kernel_size=KW or KH appropriately, and stride=1.

**Review by silvasean — CHANGES_REQUESTED** [review]

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 4 above)
Comment (ajarthurs) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:115]: Got it. Makes sense.

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 7 above)
Comment (ajarthurs) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:36]: OK

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp [hunk 8]
```diff
@@ -34,12 +34,47 @@ static SmallVector<Value, 6> bypassResultShapes(Operation *op,
   }
   // TODO: This only supports the NCHW data format. Consider other formats and lower ranks.
   if (auto conv2dNCHW = dyn_cast<tcf::ConvNCHWOp>(op)) {
+    // TODO: Replace hard-coded stride/dilation/padding constant-ops.
```
Comment (silvasean) [inline lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:37]: Maybe add a link to the PyTorch docs?

File: test/npcomp-run-mlir/conv_2d_nchw.mlir [hunk 9]
```diff
@@ -0,0 +1,40 @@
+// RUN: npcomp-run-mlir %s \
+// RUN:   -invoke conv_2d_nchw \
+// RUN:   -arg-value="dense<[[[[1.2]]]]> : tensor<1x1x1x1xf32>" \
+// RUN:   -arg-value="dense<[[[[3.4]]]]> : tensor<1x1x1x1xf32>" \
+// RUN:   -shared-libs=%npcomp_runtime_shlib 2>&1 \
+// RUN:   | FileCheck %s
+
+// RUN: npcomp-run-mlir %s \
```
Comment (silvasean) [inline test/npcomp-run-mlir/conv_2d_nchw.mlir:8]: can you test more shapes to exercise various corners of the shape calculation logic? just use dense<0> as the inputs. This is mainly to verify that we aren't reading out of bounds or allocating the wrong sizes. Outputs should be all zeros in every case.

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 10]
```diff
@@ -95,4 +95,24 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
...
 
+def TCF_ConvNCHWOp : TCF_Op<"conv_2d_nchw"> {
+  let summary = "2-D convolution";
+  let description = [{
+    Performs 2-D convolution.
+
+    The tensors have dimensions:
+    - in:     [N, C, H, W]
+    - filter: [F, C, KH, KW]
+    - result: [N, F, H, W]
```
Comment (silvasean) [inline include/npcomp/Dialect/TCF/IR/TCFOps.td:106]: this documentation is not correct.

**Review by silvasean — CHANGES_REQUESTED** [review]

File: include/npcomp/Dialect/TCF/IR/TCFOps.td (hunk 10 above)
Comment (ajarthurs) [reply include/npcomp/Dialect/TCF/IR/TCFOps.td:106]: Great catch. `result` should refer to  Hout/Wout or something besides H/W, input shape.

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 8 above)
Comment (ajarthurs) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:37]: Not sure about the PyTorch link since this op is out of PyTorch's scope. Also where's the best place to add a link? I'm thinking add a reference link to `include/npcomp/Dialect/TCF/TCFOps.td`. Do you know of a more framework-agnostic, maybe canonical, documentation for convolution operations? I'm looking around but may settle for PyTorch's docs.

Another topic related to the hard-coded stride/dilation/padding constant-ops, those are obviously placeholders and I noticed that the TCF->Linalg pass optimizes out the constant-ops, which makes sense. I don't know if it's possible to block optimization or how to declare those attributes as variables with default values.

File: test/npcomp-run-mlir/conv_2d_nchw.mlir (hunk 9 above)
Comment (ajarthurs) [reply test/npcomp-run-mlir/conv_2d_nchw.mlir:8]: Working on it...

File: lib/Conversion/TCFToLinalg/TCFToLinalg.cpp (hunk 8 above)
Comment (silvasean) [reply lib/Conversion/TCFToLinalg/TCFToLinalg.cpp:37]: Linking to the PyTorch docs is fine. This is documentation for a mathematical operator which has specific deep learning semantics and details. I don't know anywhere that is a more straightforward and canonical place to link to.

As far as the hard-coded stride/dilation/padding -- I was actually expecting the pass to only create the simplified IR in the first place. But the way you are doing it is fine as well.

File: test/npcomp-run-mlir/conv_2d_nchw.mlir [hunk 11]
```diff
@@ -1,38 +1,65 @@
...
+// RUN:   -arg-value="dense<0.0> : tensor<1x1x2x2xf32>" \
 // RUN:   -arg-value="dense<0.0> : tensor<1x1x1x1xf32>" \
-// RUN:   -arg-value="dense<0.0> : tensor<1x1x1x1xf32>" \
 // RUN:   -shared-libs=%npcomp_runtime_shlib 2>&1 \
-// RUN:   | FileCheck %s --check-prefix=ZEROS
+// RUN:   | FileCheck %s --check-prefix=TINY_SQUARE
+
+// RUN: npcomp-run-mlir %s \
+// RUN:   -invoke conv_2d_nchw \
+// RUN:   -arg-value="dense<0.0> : tensor<1x1x1024x1024xf32>" \
```
Comment (silvasean) [inline test/npcomp-run-mlir/conv_2d_nchw.mlir:31]: can we use sizes smaller than 1024? I'm worried the tests will take too long. I think that 32 is probably big enough to test what we care about.

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 12]
```diff
@@ -98,12 +98,12 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
 def TCF_ConvNCHWOp : TCF_Op<"conv_2d_nchw"> {
   let summary = "2-D convolution";
   let description = [{
-    Performs 2-D convolution.
+    Performs 2-D convolution. This op is inspired by PyTorch's Conv2d layer (https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html).
 
     The tensors have dimensions:
     - in:     [N, C, H, W]
```
Comment (silvasean) [inline include/npcomp/Dialect/TCF/IR/TCFOps.td:104]: use the PyTorch dimension terminology since we are referencing their docs Cin, Cout, Hin, Win.

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 13]
```diff
@@ -98,12 +98,12 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
...
     - filter: [F, C, KH, KW]
-    - result: [N, F, H, W]
+    - result: [N, F, Hout, Wout]
 
     The tensors must met the following conditions; otherwise, this op aborts the program.
```
Comment (silvasean) [inline include/npcomp/Dialect/TCF/IR/TCFOps.td:108]: typo: must meet

File: include/npcomp/Dialect/TCF/IR/TCFOps.td [hunk 14]
```diff
@@ -95,4 +95,24 @@ def TCF_MatmulOp : TCF_Op<"matmul"> {
...
+  let description = [{
+    Performs 2-D convolution. This op is inspired by PyTorch's Conv2d layer (https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html).
+
+    The tensors have dimensions:
+    - in:     [N, C, H, W]
+    - filter: [F, C, KH, KW]
+    - result: [N, F, Hout, Wout]
+
+    The tensors must met the following conditions; otherwise, this op aborts the program.
+    - H is greater than or equal to KH
```
Comment (silvasean) [inline include/npcomp/Dialect/TCF/IR/TCFOps.td:109]: also mention that this op aborts the program if the above tensor dimensions don't match when we use the same symbol for two dimensions.

**Review by silvasean — APPROVED** [review]

**Comment by silvasean** [comment]
Let me know when you want me to push this for you!

**Comment by ajarthurs** [comment]
Let me know if you want `filter` renamed to `kernel` and I'll cherry-pick another commit. It makes no difference to me one way or the other.

**Review by silvasean — APPROVED** [review]
